File: swim-java/swim-core/swim.uri/src/main/java/swim/uri/UriAuthority.java
// Copyright 2015-2022 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.uri;

import java.io.IOException;
import java.util.Objects;
import swim.annotations.Nullable;
import swim.annotations.Public;
import swim.annotations.Since;
import swim.codec.Diagnostic;
import swim.codec.Input;
import swim.codec.Parse;
import swim.codec.ParseException;
import swim.codec.StringInput;
import swim.decl.Marshal;
import swim.decl.Unmarshal;
import swim.util.Assume;
import swim.util.CacheMap;
import swim.util.LruCacheMap;
import swim.util.Murmur3;
import swim.util.Notation;
import swim.util.ToSource;
import swim.util.ToString;

@Public
@Since("5.0")
public final class UriAuthority extends UriPart implements Comparable<UriAuthority>, ToSource, ToString {

  final UriUser user;

  final UriHost host;

  final UriPort port;

  transient @Nullable String string;

  UriAuthority(UriUser user, UriHost host, UriPort port) {
    this.user = user;
    this.host = host;
    this.port = port;
  }

  public boolean isDefined() {
    return this.user.isDefined() || this.host.isDefined() || this.port.isDefined();
  }

  public UriUser user() {
    return this.user;
  }

  @SuppressWarnings("ReferenceEquality")
  public UriAuthority withUser(UriUser user) {
    if (user == this.user) {
      return this;
    }
    return UriAuthority.of(user, this.host, this.port);
  }

  public String userPart() {
    return this.user.toString();
  }

  public UriAuthority withUserPart(String userPart) {
    try {
      return this.withUser(UriUser.parse(userPart).getNonNull());
    } catch (ParseException cause) {
      throw new IllegalArgumentException(Notation.of("malformed user part: ")
                                                 .appendSource(userPart)
                                                 .toString(), cause);
    }
  }

  public UriAuthority withUserNamePass(@Nullable String userName,
                                       @Nullable String userPass) {
    return this.withUser(UriUser.namePass(userName, userPass));
  }

  public @Nullable String userName() {
    return this.user.name();
  }

  public UriAuthority withUserName(@Nullable String userName) {
    return this.withUser(this.user.withName(userName));
  }

  public @Nullable String userPass() {
    return this.user.pass();
  }

  public UriAuthority withUserPass(@Nullable String userPass) {
    return this.withUser(this.user.withPass(userPass));
  }

  public UriHost host() {
    return this.host;
  }

  @SuppressWarnings("ReferenceEquality")
  public UriAuthority withHost(UriHost host) {
    if (host == this.host) {
      return this;
    }
    return UriAuthority.of(this.user, host, this.port);
  }

  public String hostPart() {
    return this.host.toString();
  }

  public UriAuthority withHostPart(String hostPart) {
    try {
      return this.withHost(UriHost.parse(hostPart).getNonNull());
    } catch (ParseException cause) {
      throw new IllegalArgumentException(Notation.of("malformed host part: ")
                                                 .appendSource(hostPart)
                                                 .toString(), cause);
    }
  }

  public String hostAddress() {
    return this.host.address();
  }

  public @Nullable String hostName() {
    return this.host.name();
  }

  public UriAuthority withHostName(@Nullable String hostName) {
    return this.withHost(UriHost.name(hostName));
  }

  public @Nullable String hostIPv4() {
    return this.host.ipv4();
  }

  public UriAuthority withHostIPv4(@Nullable String hostIPv4) {
    return this.withHost(UriHost.ipv4(hostIPv4));
  }

  public @Nullable String hostIPv6() {
    return this.host.ipv6();
  }

  public UriAuthority withHostIPv6(@Nullable String hostIPv6) {
    return this.withHost(UriHost.ipv6(hostIPv6));
  }

  public UriPort port() {
    return this.port;
  }

  @SuppressWarnings("ReferenceEquality")
  public UriAuthority withPort(UriPort port) {
    if (port == this.port) {
      return this;
    }
    return UriAuthority.of(this.user, this.host, port);
  }

  public String portPart() {
    return this.port.toString();
  }

  public UriAuthority withPortPart(String portPart) {
    try {
      return this.withPort(UriPort.parse(portPart).getNonNull());
    } catch (ParseException cause) {
      throw new IllegalArgumentException(Notation.of("malformed port part: ")
                                                 .appendSource(portPart)
                                                 .toString(), cause);
    }
  }

  public int portNumber() {
    return this.port.number();
  }

  public UriAuthority withPortNumber(int portNumber) {
    return this.withPort(UriPort.number(portNumber));
  }

  @Override
  public int compareTo(UriAuthority that) {
    return this.toString().compareTo(that.toString());
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) {
      return true;
    } else if (other instanceof UriAuthority that) {
      return this.toString().equals(that.toString());
    }
    return false;
  }

  @Override
  public int hashCode() {
    return Murmur3.seed(this.toString());
  }

  @Override
  public void writeSource(Appendable output) {
    final Notation notation = Notation.from(output);
    if (this.user.isDefined()) {
      notation.beginInvoke("UriAuthority", "of")
              .appendArgument(this.user)
              .appendArgument(this.host)
              .appendArgument(this.port)
              .endInvoke();
    } else if (this.host.isDefined()) {
      if (this.host.name() != null) {
        notation.beginInvoke("UriAuthority", "hostName")
                .appendArgument(this.host.name())
                .endInvoke();
      } else if (this.host.ipv4() != null) {
        notation.beginInvoke("UriAuthority", "hostIPv4")
                .appendArgument(this.host.ipv4())
                .endInvoke();
      } else if (this.host.ipv6() != null) {
        notation.beginInvoke("UriAuthority", "hostIPv6")
                .appendArgument(this.host.ipv6())
                .endInvoke();
      } else {
        notation.beginInvoke("UriAuthority", "host")
                .appendArgument(this.host)
                .endInvoke();
      }
      if (this.port.isDefined()) {
        notation.beginInvoke("withPortNumber")
                .appendArgument(this.portNumber())
                .endInvoke();
      }
    } else if (this.port.isDefined()) {
      notation.beginInvoke("UriAuthority", "portNumber")
              .appendArgument(this.portNumber())
              .endInvoke();
    } else {
      notation.beginInvoke("UriAuthority", "undefined").endInvoke();
    }
  }

  @Override
  public void writeString(Appendable output) throws IOException {
    if (this.string != null) {
      output.append(this.string);
      return;
    }
    if (this.user.isDefined()) {
      this.user.writeString(output);
      output.append('@');
    }
    this.host.writeString(output);
    if (this.port.isDefined()) {
      output.append(':');
      this.port.writeString(output);
    }
  }

  @Marshal
  @Override
  public String toString() {
    if (this.string == null) {
      this.string = this.toString(null);
    }
    return this.string;
  }

  static final UriAuthority UNDEFINED = new UriAuthority(UriUser.undefined(),
                                                         UriHost.undefined(),
                                                         UriPort.undefined());

  public static UriAuthority undefined() {
    return UNDEFINED;
  }

  public static UriAuthority of(@Nullable UriUser user,
                                @Nullable UriHost host,
                                @Nullable UriPort port) {
    if (user == null) {
      user = UriUser.undefined();
    }
    if (host == null) {
      host = UriHost.undefined();
    }
    if (port == null) {
      port = UriPort.undefined();
    }
    if (!user.isDefined() && !host.isDefined() && !port.isDefined()) {
      return UriAuthority.undefined();
    }
    return new UriAuthority(user, host, port);
  }

  public static UriAuthority user(@Nullable UriUser user) {
    return UriAuthority.of(user, null, null);
  }

  public static UriAuthority userNamePass(@Nullable String userName,
                                          @Nullable String userPass) {
    return UriAuthority.user(UriUser.namePass(userName, userPass));
  }

  public static UriAuthority userName(@Nullable String userName) {
    return UriAuthority.user(UriUser.name(userName));
  }

  public static UriAuthority host(@Nullable UriHost host) {
    return UriAuthority.of(null, host, null);
  }

  public static UriAuthority hostName(@Nullable String hostName) {
    return UriAuthority.host(UriHost.name(hostName));
  }

  public static UriAuthority hostIPv4(@Nullable String hostIPv4) {
    return UriAuthority.host(UriHost.ipv4(hostIPv4));
  }

  public static UriAuthority hostIPv6(@Nullable String hostIPv6) {
    return UriAuthority.host(UriHost.ipv6(hostIPv6));
  }

  public static UriAuthority port(@Nullable UriPort port) {
    return UriAuthority.of(null, null, port);
  }

  public static UriAuthority portNumber(int portNumber) {
    return UriAuthority.port(UriPort.number(portNumber));
  }

  @Unmarshal
  public static @Nullable UriAuthority from(String value) {
    return UriAuthority.parse(value).getOr(null);
  }

  public static Parse<UriAuthority> parse(Input input) {
    return ParseUriAuthority.parse(input, null, null, null, 1);
  }

  public static Parse<UriAuthority> parse(String part) {
    Objects.requireNonNull(part);
    final CacheMap<String, Parse<UriAuthority>> cache = UriAuthority.cache();
    Parse<UriAuthority> parseAuthority = cache.get(part);
    if (parseAuthority == null) {
      final StringInput input = new StringInput(part);
      parseAuthority = UriAuthority.parse(input).complete(input);
      if (parseAuthority.isDone()) {
        // Don't cache user info.
        if (!parseAuthority.getNonNullUnchecked().user.isDefined()) {
          parseAuthority = cache.put(part, parseAuthority);
        }
      }
    }
    return parseAuthority;
  }

  static final ThreadLocal<CacheMap<String, Parse<UriAuthority>>> CACHE =
      new ThreadLocal<CacheMap<String, Parse<UriAuthority>>>();

  static CacheMap<String, Parse<UriAuthority>> cache() {
    CacheMap<String, Parse<UriAuthority>> cache = CACHE.get();
    if (cache == null) {
      int cacheSize;
      try {
        cacheSize = Integer.parseInt(System.getProperty("swim.uri.authority.cache.size"));
      } catch (NumberFormatException cause) {
        cacheSize = 128;
      }
      cache = new LruCacheMap<String, Parse<UriAuthority>>(cacheSize);
      CACHE.set(cache);
    }
    return cache;
  }

}

final class ParseUriAuthority extends Parse<UriAuthority> {

  final @Nullable Parse<UriUser> parseUser;
  final @Nullable Parse<UriHost> parseHost;
  final @Nullable Parse<UriPort> parsePort;
  final int step;

  ParseUriAuthority(@Nullable Parse<UriUser> parseUser,
                    @Nullable Parse<UriHost> parseHost,
                    @Nullable Parse<UriPort> parsePort, int step) {
    this.parseUser = parseUser;
    this.parseHost = parseHost;
    this.parsePort = parsePort;
    this.step = step;
  }

  @Override
  public Parse<UriAuthority> consume(Input input) {
    return ParseUriAuthority.parse(input, this.parseUser, this.parseHost,
                                   this.parsePort, this.step);
  }

  static Parse<UriAuthority> parse(Input input, @Nullable Parse<UriUser> parseUser,
                                   @Nullable Parse<UriHost> parseHost,
                                   @Nullable Parse<UriPort> parsePort, int step) {
    int c = 0;
    if (step == 1) {
      if (input.isCont()) {
        final Input lookahead = input.clone();
        while (lookahead.isCont() && (c = lookahead.head()) != '@' && c != '/') {
          lookahead.step();
        }
        if (lookahead.isCont() && c == '@') {
          step = 2;
        } else {
          step = 3;
        }
      } else if (input.isDone()) {
        step = 3;
      }
    }
    if (step == 2) {
      if (parseUser == null) {
        parseUser = UriUser.parse(input);
      } else {
        parseUser = parseUser.consume(input);
      }
      if (parseUser.isDone()) {
        if (input.isCont() && input.head() == '@') {
          input.step();
          step = 3;
        } else if (input.isReady()) {
          return Parse.error(Diagnostic.expected('@', input));
        }
      } else if (parseUser.isError()) {
        return parseUser.asError();
      }
    }
    if (step == 3) {
      if (parseHost == null) {
        parseHost = UriHost.parse(input);
      } else {
        parseHost = parseHost.consume(input);
      }
      if (parseHost.isDone()) {
        if (input.isCont() && input.head() == ':') {
          input.step();
          step = 4;
        } else if (input.isReady()) {
          return Parse.done(UriAuthority.of(parseUser != null ? parseUser.getUnchecked() : null,
                                            parseHost.getUnchecked(),
                                            null));
        }
      } else if (parseHost.isError()) {
        return parseHost.asError();
      }
    }
    if (step == 4) {
      if (parsePort == null) {
        parsePort = UriPort.parse(input);
      } else {
        parsePort = parsePort.consume(input);
      }
      if (parsePort.isDone()) {
        return Parse.done(UriAuthority.of(parseUser != null ? parseUser.getUnchecked() : null,
                                          Assume.nonNull(parseHost).getUnchecked(),
                                          parsePort.getUnchecked()));
      } else if (parsePort.isError()) {
        return parsePort.asError();
      }
    }
    if (input.isError()) {
      return Parse.error(input.getError());
    }
    return new ParseUriAuthority(parseUser, parseHost, parsePort, step);
  }

}

File: swim-js/swim-core/swim-uri/main/UriAuthority.ts
// Copyright 2015-2023 Nstream, inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type {Uninitable} from "@swim/util";
import type {Mutable} from "@swim/util";
import {Lazy} from "@swim/util";
import type {HashCode} from "@swim/util";
import type {Compare} from "@swim/util";
import {Strings} from "@swim/util";
import {Objects} from "@swim/util";
import {Diagnostic} from "@swim/codec";
import type {Input} from "@swim/codec";
import type {Output} from "@swim/codec";
import {Parser} from "@swim/codec";
import type {Debug} from "@swim/codec";
import type {Display} from "@swim/codec";
import {Format} from "@swim/codec";
import {Unicode} from "@swim/codec";
import type {UriUserLike} from "./UriUser";
import type {UriUserInit} from "./UriUser";
import {UriUser} from "./"; // forward import
import type {UriHostLike} from "./UriHost";
import {UriHost} from "./"; // forward import
import type {UriPortLike} from "./UriPort";
import {UriPort} from "./"; // forward import
import {UriUserParser} from "./"; // forward import
import {UriHostParser} from "./"; // forward import
import {UriPortParser} from "./"; // forward import

/** @public */
export type UriAuthorityLike = UriAuthority | UriAuthorityInit | string;

/** @public */
export const UriAuthorityLike = {
  [Symbol.hasInstance](instance: unknown): instance is UriAuthorityLike {
    return instance instanceof UriAuthority
        || UriAuthorityInit[Symbol.hasInstance](instance)
        || typeof instance === "string";
  },
};

/** @public */
export interface UriAuthorityInit extends UriUserInit {
  /** @internal */
  readonly typeid?: "UriAuthorityInit" | "UriInit";
  user?: UriUserLike;
  host?: UriHostLike;
  port?: UriPortLike;
}

/** @public */
export const UriAuthorityInit = {
  [Symbol.hasInstance](instance: unknown): instance is UriAuthorityInit {
    return Objects.hasAnyKey<UriAuthorityInit>(instance, "user", "host", "port");
  },
};

/** @public */
export class UriAuthority implements HashCode, Compare, Debug, Display {
  /** @internal */
  constructor(user: UriUser, host: UriHost, port: UriPort) {
    this.user = user;
    this.host = host;
    this.port = port;
    this.hashValue = void 0;
    this.stringValue = void 0;
  }

  /** @internal */
  declare readonly typeid?: "UriAuthority";

  likeType?(like: UriAuthorityInit | string): void;

  isDefined(): boolean {
    return this.user.isDefined() || this.host.isDefined() || this.port.isDefined();
  }

  readonly user: UriUser;

  withUser(user: UriUserLike): UriAuthority {
    user = UriUser.fromLike(user);
    if (user === this.user) {
      return this;
    }
    return this.copy(user, this.host, this.port);
  }

  get userPart(): string {
    return this.user.toString();
  }

  withUserPart(userPart: string): UriAuthority {
    return this.withUser(UriUser.parse(userPart));
  }

  get username(): string | undefined {
    return this.user.username;
  }

  withUsername(username: string | undefined, password?: string): UriAuthority {
    if (arguments.length === 1) {
      return this.withUser(this.user.withUsername(username));
    } else if (arguments.length === 2) {
      return this.withUser(UriUser.create(username, password));
    }
    throw new Error(arguments.toString());
  }

  get password(): string | undefined {
    return this.user.password;
  }

  withPassword(password: string | undefined): UriAuthority {
    return this.withUser(this.user.withPassword(password));
  }

  readonly host: UriHost;

  withHost(host: UriHostLike): UriAuthority {
    host = UriHost.fromLike(host);
    if (host === this.host) {
      return this;
    }
    return this.copy(this.user, host, this.port);
  }

  get hostPart(): string {
    return this.host.toString();
  }

  withHostPart(hostPart: string): UriAuthority {
    return this.withHost(UriHost.parse(hostPart));
  }

  get hostAddress(): string {
    return this.host.address;
  }

  get hostName(): string | undefined {
    return this.host.name;
  }

  withHostName(hostName: string): UriAuthority {
    return this.withHost(UriHost.hostname(hostName));
  }

  get hostIPv4(): string | undefined {
    return this.host.ipv4;
  }

  withHostIPv4(hostIPv4: string): UriAuthority {
    return this.withHost(UriHost.ipv4(hostIPv4));
  }

  get hostIPv6(): string | undefined {
    return this.host.ipv6;
  }

  withHostIPv6(hostIPv5: string): UriAuthority {
    return this.withHost(UriHost.ipv6(hostIPv5));
  }

  readonly port: UriPort;

  withPort(port: UriPortLike): UriAuthority {
    port = UriPort.fromLike(port);
    if (port === this.port) {
      return this;
    }
    return this.copy(this.user, this.host, port);
  }

  get portPart(): string {
    return this.port.toString();
  }

  withPortPart(portPart: string): UriAuthority {
    return this.withPort(UriPort.parse(portPart));
  }

  get portNumber(): number {
    return this.port.number;
  }

  withPortNumber(portNumber: number): UriAuthority {
    return this.withPort(UriPort.create(portNumber));
  }

  protected copy(user: UriUser, host: UriHost, port: UriPort): UriAuthority {
    return UriAuthority.create(user, host, port);
  }

  toLike(authority?: {username?: string, password?: string, host?: string, port?: number}):
      {username?: string, password?: string, host?: string, port?: number} | undefined {
    if (this.isDefined()) {
      if (authority === void 0) {
        authority = {};
      }
      this.user.toLike(authority);
      if (this.host.isDefined()) {
        authority.host = this.host.toLike();
      }
      if (this.port.isDefined()) {
        authority.port = this.port.toLike();
      }
    }
    return authority;
  }

  /** @override */
  compareTo(that: unknown): number {
    if (that instanceof UriAuthority) {
      return this.toString().localeCompare(that.toString());
    }
    return NaN;
  }

  /** @override */
  equals(that: unknown): boolean {
    if (this === that) {
      return true;
    } else if (that instanceof UriAuthority) {
      return this.toString() === that.toString();
    }
    return false;
  }

  /** @internal */
  readonly hashValue: number | undefined;

  /** @override */
  hashCode(): number {
    let hashValue = this.hashValue;
    if (hashValue === void 0) {
      hashValue = Strings.hash(this.toString());
      (this as Mutable<this>).hashValue = hashValue;
    }
    return hashValue;
  }

  /** @override */
  debug<T>(output: Output<T>): Output<T> {
    output = output.write("UriAuthority").write(46/*'.'*/);
    if (this.isDefined()) {
      output = output.write("parse").write(40/*'('*/).write(34/*'"'*/)
                     .display(this).write(34/*'"'*/).write(41/*')'*/);
    } else {
      output = output.write("undefined").write(40/*'('*/).write(41/*')'*/);
    }
    return output;
  }

  /** @override */
  display<T>(output: Output<T>): Output<T> {
    const stringValue = this.stringValue;
    if (stringValue !== void 0) {
      output = output.write(stringValue);
    } else {
      const user = this.user;
      if (user.isDefined()) {
        output = output.display(user).write(64/*'@'*/);
      }
      output = output.display(this.host);
      const port = this.port;
      if (port.isDefined()) {
        output = output.write(58/*':'*/).display(port);
      }
    }
    return output;
  }

  /** @internal */
  readonly stringValue: string | undefined;

  /** @override */
  toString(): string {
    let stringValue = this.stringValue;
    if (stringValue === void 0) {
      stringValue = Format.display(this);
      (this as Mutable<this>).stringValue = stringValue;
    }
    return stringValue;
  }

  @Lazy
  static undefined(): UriAuthority {
    return new UriAuthority(UriUser.undefined(), UriHost.undefined(), UriPort.undefined());
  }

  static create(user: UriUser = UriUser.undefined(),
                host: UriHost = UriHost.undefined(),
                port: UriPort = UriPort.undefined()): UriAuthority {
    if (!user.isDefined() && !host.isDefined() && !port.isDefined()) {
      return UriAuthority.undefined();
    }
    return new UriAuthority(user, host, port);
  }

  static fromLike<T extends UriAuthorityLike | null | undefined>(value: T): UriAuthority | Uninitable<T> {
    if (value === void 0 || value === null || value instanceof UriAuthority) {
      return value as UriAuthority | Uninitable<T>;
    } else if (typeof value === "object") {
      return UriAuthority.fromInit(value);
    } else if (typeof value === "string") {
      return UriAuthority.parse(value);
    }
    throw new TypeError("" + value);
  }

  static fromInit(init: UriAuthorityInit): UriAuthority {
    const user = UriUser.fromLike(init.user);
    const host = UriHost.fromLike(init.host);
    const port = UriPort.fromLike(init.port);
    return this.create(user, host, port);
  }

  static user(user: UriUserLike): UriAuthority {
    user = UriUser.fromLike(user);
    return this.create(user, void 0, void 0);
  }

  static userPart(userPart: string): UriAuthority {
    const user = UriUser.parse(userPart);
    return this.create(user, void 0, void 0);
  }

  static username(username: string, password?: string | undefined): UriAuthority {
    const user = UriUser.create(username, password);
    return this.create(user, void 0, void 0);
  }

  static password(password: string): UriAuthority {
    const user = UriUser.create("", password);
    return this.create(user, void 0, void 0);
  }

  static host(host: UriHostLike): UriAuthority {
    host = UriHost.fromLike(host);
    return this.create(void 0, host, void 0);
  }

  static hostPart(hostPart: string): UriAuthority {
    const host = UriHost.parse(hostPart);
    return this.create(void 0, host, void 0);
  }

  static hostName(hostName: string): UriAuthority {
    const host = UriHost.hostname(hostName);
    return this.create(void 0, host, void 0);
  }

  static hostIPv4(hostIPv4: string): UriAuthority {
    const host = UriHost.ipv4(hostIPv4);
    return this.create(void 0, host, void 0);
  }

  static hostIPv6(hostIPv6: string): UriAuthority {
    const host = UriHost.ipv6(hostIPv6);
    return this.create(void 0, host, void 0);
  }

  static port(port: UriPortLike): UriAuthority {
    port = UriPort.fromLike(port);
    return this.create(void 0, void 0, port);
  }

  static portPart(portPart: string): UriAuthority {
    const port = UriPort.parse(portPart);
    return this.create(void 0, void 0, port);
  }

  static portNumber(portNumber: number): UriAuthority {
    const port = UriPort.create(portNumber);
    return this.create(void 0, void 0, port);
  }

  static parse(input: Input): Parser<UriAuthority>;
  static parse(string: string): UriAuthority;
  static parse(string: Input | string): Parser<UriAuthority> | UriAuthority {
    const input = typeof string === "string" ? Unicode.stringInput(string) : string;
    let parser = UriAuthorityParser.parse(input);
    if (typeof string === "string" && input.isCont() && !parser.isError()) {
      parser = Parser.error(Diagnostic.unexpected(input));
    }
    return typeof string === "string" ? parser.bind() : parser;
  }
}

/** @internal */
export class UriAuthorityParser extends Parser<UriAuthority> {
  private readonly userParser: Parser<UriUser> | undefined;
  private readonly hostParser: Parser<UriHost> | undefined;
  private readonly portParser: Parser<UriPort> | undefined;
  private readonly step: number | undefined;

  constructor(userParser?: Parser<UriUser>, hostParser?: Parser<UriHost>,
              portParser?: Parser<UriPort>, step?: number) {
    super();
    this.userParser = userParser;
    this.hostParser = hostParser;
    this.portParser = portParser;
    this.step = step;
  }

  override feed(input: Input): Parser<UriAuthority> {
    return UriAuthorityParser.parse(input, this.userParser, this.hostParser,
                                    this.portParser, this.step);
  }

  static parse(input: Input, userParser?: Parser<UriUser>, hostParser?: Parser<UriHost>,
               portParser?: Parser<UriPort>, step: number = 1): Parser<UriAuthority> {
    let c = 0;
    if (step === 1) {
      if (input.isCont()) {
        const look = input.clone();
        while (look.isCont() && (c = look.head(), c !== 64/*'@'*/ && c !== 47/*'/'*/)) {
          look.step();
        }
        if (look.isCont() && c === 64/*'@'*/) {
          step = 2;
        } else {
          step = 3;
        }
      } else if (input.isDone()) {
        step = 3;
      }
    }
    if (step === 2) {
      if (userParser === void 0) {
        userParser = UriUserParser.parse(input);
      } else {
        userParser = userParser.feed(input);
      }
      if (userParser.isDone()) {
        if (input.isCont() && input.head() === 64/*'@'*/) {
          input = input.step();
          step = 3;
        } else if (!input.isEmpty()) {
          return Parser.error(Diagnostic.expected(64/*'@'*/, input));
        }
      } else if (userParser.isError()) {
        return userParser.asError();
      }
    }
    if (step === 3) {
      if (hostParser === void 0) {
        hostParser = UriHostParser.parse(input);
      } else {
        hostParser = hostParser.feed(input);
      }
      if (hostParser.isDone()) {
        if (input.isCont() && input.head() === 58/*':'*/) {
          input = input.step();
          step = 4;
        } else if (!input.isEmpty()) {
          return Parser.done(UriAuthority.create(userParser !== void 0 ? userParser.bind() : void 0,
                                                 hostParser.bind()));
        }
      } else if (hostParser.isError()) {
        return hostParser.asError();
      }
    }
    if (step === 4) {
      if (portParser === void 0) {
        portParser = UriPortParser.parse(input);
      } else {
        portParser = portParser.feed(input);
      }
      if (portParser.isDone()) {
        return Parser.done(UriAuthority.create(userParser !== void 0 ? userParser.bind() : void 0,
                                               hostParser!.bind(),
                                               portParser.bind()));
      } else if (portParser.isError()) {
        return portParser.asError();
      }
    }
    return new UriAuthorityParser(userParser, hostParser, portParser, step);
  }
}
